File: client/src/test/scala/stasis/test/specs/unit/client/mocks/MockCompression.scala
package stasis.test.specs.unit.client.mocks

import java.nio.file.Path
import java.util.concurrent.atomic.AtomicInteger

import org.apache.pekko.NotUsed
import org.apache.pekko.stream.scaladsl.Flow
import org.apache.pekko.util.ByteString

import stasis.client.compression.Compression
import stasis.client.compression.Decoder
import stasis.client.compression.Encoder
import stasis.client.model.SourceEntity
import stasis.client.model.TargetEntity
import stasis.test.specs.unit.client.mocks.MockCompression.Statistic

class MockCompression() extends Compression with Encoder with Decoder {
  override val name: String = "mock"

  override val defaultCompression: Encoder with Decoder = this

  override val disabledExtensions: Set[String] = Set("test")

  override def algorithmFor(entity: Path): String = "mock"
  override def encoderFor(entity: SourceEntity): Encoder = this
  override def decoderFor(entity: TargetEntity): Decoder = this

  private val stats: Map[Statistic, AtomicInteger] = Map(
    Statistic.Compressed -> new AtomicInteger(0),
    Statistic.Decompressed -> new AtomicInteger(0)
  )

  override def compress: Flow[ByteString, ByteString, NotUsed] =
    Flow[ByteString]
      .map { _ =>
        stats(Statistic.Compressed).incrementAndGet()
        ByteString("compressed")
      }

  override def decompress: Flow[ByteString, ByteString, NotUsed] =
    Flow[ByteString]
      .map { _ =>
        stats(Statistic.Decompressed).incrementAndGet()
        ByteString("decompressed")
      }

  def statistics: Map[Statistic, Int] = stats.view.mapValues(_.get()).toMap
}

object MockCompression {
  def apply(): MockCompression = new MockCompression()

  sealed trait Statistic
  object Statistic {
    case object Compressed extends Statistic
    case object Decompressed extends Statistic
  }
}

File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/mocks/MockCompression.kt
package stasis.test.client_android.lib.mocks

import okio.Source
import stasis.client_android.lib.compression.Compression
import stasis.client_android.lib.compression.Compressor
import stasis.client_android.lib.compression.Decoder
import stasis.client_android.lib.compression.Encoder
import stasis.client_android.lib.model.SourceEntity
import stasis.client_android.lib.model.TargetEntity
import java.util.concurrent.atomic.AtomicInteger

open class MockCompression : Compression, Compressor {
    private val stats: Map<Statistic, AtomicInteger> = mapOf(
        Statistic.Compressed to AtomicInteger(0),
        Statistic.Decompressed to AtomicInteger(0)
    )

    override fun defaultCompression(): Compressor = this

    override fun disabledExtensions(): Set<String> = setOf("test")

    override fun name(): String = "mock"

    override fun encoderFor(entity: SourceEntity): Encoder = this

    override fun decoderFor(entity: TargetEntity): Decoder = this

    override fun compress(source: Source): Source {
        stats[Statistic.Compressed]?.getAndIncrement()
        return source
    }

    override fun decompress(source: Source): Source {
        stats[Statistic.Decompressed]?.getAndIncrement()
        return source
    }

    val statistics: Map<Statistic, Int>
        get() = stats.mapValues { it.value.get() }

    sealed class Statistic {
        object Compressed : Statistic()
        object Decompressed : Statistic()
    }
}
